### PR TITLE
ISPN-10458 Address comments on ISPN-9723

### PR DESCRIPTION
**Pull request by wburns**
https://issues.jboss.org/browse/ISPN-10458

### CODE REVIEW

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java [hunk 1]
```diff
@@ -211,22 +213,20 @@ public boolean containsKey(Object k) {
...
+         return CompletableFutures.completedNull();
       }
-      return CompletableFutures.completedNull();
+      ByRef<CompletionStage<Void>> byRef = new ByRef<>(CompletableFutures.completedNull());
+      entries.computeIfPresent(key, (o, entry) -> {
+         // Note this is non blocking and we ignore the return value
+         // - we don't need an orderer as it is handled in OrderedClusteringDependentLogic
+         // - we don't need eviction manager either as it is handled in NotifyHelper
+         byRef.set(handleEviction(entry, null, passivator.running(), null, this,
+               null));
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: Single line?

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java [hunk 2]
```diff
@@ -498,32 +498,31 @@ public void delete(K key, InternalCacheEntry<K, V> value, RemovalCause cause) {
...
     * @param <K> key type of the entry
     * @param <V> value type of the entry
     * @return stage that when complete all of the eviction logic is complete
     */
-   public static <K, V> CompletionStage<Void> handleEviction(K key, InternalCacheEntry<K, V> value,
-         NonBlockingOrderer orderer, PassivationManager passivator, EvictionManager<K, V> evictionManager,
-         DataContainer<K, V> dataContainer) {
+   public static <K, V> CompletionStage<Void> handleEviction(InternalCacheEntry<K, V> entry, NonBlockingOrderer orderer,
+         PassivationManager passivator, EvictionManager<K, V> evictionManager, DataContainer<K, V> dataContainer,
+         CompletionStage<Void> writeDelay) {
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: Is `writeDelay` the best name for this as it's a future not a duration. How about something like `evictionSyncFuture`?

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java (hunk 1 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: Sure.

**Comment by danberindei** [comment]
@wburns https://github.com/infinispan/infinispan/pull/7168#pullrequestreview-271762541 :)

**Comment by wburns** [comment]
@danberindei 
> One thing that bothers me in a lot of tests is that you're using eventually* because notification/passivation/activation are non-blocking, but IMO even if they are non-blocking they should still prevent the read/write that triggered them from finishing.

So technically these are protected by the NonBlockingOrderer so atomicity is not lost. I figured it would be nice to improve throughput a bit by not having it block as we are just waiting on the store anyways. Unfortunately it makes usability for the product easier but tests are a bit more annoying. I thought it was a fair tradeoff

**Comment by danberindei** [comment]
> @danberindei
> 
> > One thing that bothers me in a lot of tests is that you're using eventually* because notification/passivation/activation are non-blocking, but IMO even if they are non-blocking they should still prevent the read/write that triggered them from finishing.
> 
> So technically these are protected by the NonBlockingOrderer so atomicity is not lost. I figured it would be nice to improve throughput a bit by not having it block as we are just waiting on the store anyways. Unfortunately it makes usability for the product easier but tests are a bit more annoying. I thought it was a fair tradeoff

I see... I thought you'd have to wait for the store to finish, so that the `PassivationPersistenceManager` map size would be limited to the number of concurrent cache operations, otherwise the map could grow indefinitely.

**Comment by wburns** [comment]
> I see... I thought you'd have to wait for the store to finish, so that the `PassivationPersistenceManager` map size would be limited to the number of concurrent cache operations, otherwise the map could grow indefinitely.

Technically waiting would help if a single thread was invoking `put` in a tight loop, but it wouldn't help `putAsync` or others. But doing `put` in a tight loop you would already be buying yourself breathing room as the user would be stuck waiting on other things besides the store, such as network replication - so it isn't as big of a deal as `putAsync` which is unaffected.

**Comment by wburns** [comment]
> Overall looks good, just a couple of very minor things. Also, I'm happy keeping this as three commits, but could you make the inner message the top-level commit message. For example: "ISPN-10458 Fixed window of losing entry during activation and passivation".

Our guideline is to have the JIRA name as the first line.

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java (hunk 2 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: Well technically it is used solely for the purpose of delaying the operation and we have called similar things this in the async interceptor stack.

**Comment by danberindei** [comment]
> Technically waiting would help if a single thread was invoking put in a tight loop, but it wouldn't help putAsync or others. But doing put in a tight loop you would already be buying yourself breathing room as the user would be stuck waiting on other things besides the store, such as network replication - so it isn't as big of a deal as putAsync which is unaffected.

Even for `putAsync` in a loop we can warn the users that they should set a limit on the number of concurrent operations, otherwise their cache could use unlimited memory. (And eventually I espect us to implement a mechanism to limit the number of concurrent operations, blocking or not.) Without delaying the operation end, I'm not sure there's anything users could do to limit the number of entries in the `PassivationPersistenceManager` map.

> Our guideline is to have the JIRA name as the first line.

@wburns IMO that only applies if you have a single commit for the issue. Most of the time one sees only show the commit subject (i.e. the first line), so having 3 commits with the same subject doesn't look good.

**Comment by wburns** [comment]
Everything should be accounted for except renaming commits and 

https://github.com/infinispan/infinispan/pull/7168/files/689930be94e2c3d51f4d19b8f92fbefa6af537cf#diff-d70f5361c5448faaa6e2d954a02b42bb
https://github.com/infinispan/infinispan/pull/7168/files/689930be94e2c3d51f4d19b8f92fbefa6af537cf#diff-734dc64b1cd5e29940a872f024271249

**Comment by wburns** [comment]
Just the two comments remain. I think they should be able to fixed separately. This PR itself fixes a bug that causes some tests to fail intermittently, where as the others are either an extreme edge case or something that needs to be investigated further.

**Comment by wburns** [comment]
Two other issues we will want to log

https://github.com/infinispan/infinispan/pull/7168#discussion_r313531023
https://github.com/infinispan/infinispan/pull/7168#discussion_r313452915

**Comment by ryanemerson** [comment]
> Our guideline is to have the JIRA name as the first line.

I thought our guideline was just for the JIRA ID to be in the top line. I don't think we ever stipulated that the top line message always has to be the same as the Jira title. If we did, we certainly don't follow that rule. For example: https://github.com/infinispan/infinispan/compare/e95d5a21826cd8fe3bb242a64fee47ac0c35a66c...9975bdc83b8e3d31a630f3b1ddc85570f6ab3b14

**Comment by wburns** [comment]
Oh I agree it hasn't been closely followed by everyone.
https://infinispan.org/docs/stable/contributing/contributing.html#comments

**Comment by ryanemerson** [comment]
@wburns NPE on `TxStoreTest.testTxRollbackOnStoreException` still. The commits need rebasing etc as well

**Comment by wburns** [comment]
ah didn't notice it, fixing.

**Comment by wburns** [comment]
NPE fixed.

File: core/src/test/java/org/infinispan/eviction/impl/EvictionWithConcurrentOperationsTest.java [hunk 3]
```diff
@@ -374,6 +385,87 @@ public void testScenario7() throws Exception {
       assertInMemory(key1, "v2");
    }
 
+   // Data is written to container but before releasing orderer the entry is evicted. In this case the entry
+   // should be passivated to ensure data is still around
+   public void testEvictionDuringWrite() throws InterruptedException, ExecutionException, TimeoutException {
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/eviction/impl/EvictionWithConcurrentOperationsTest.java]: Maybe we also need to test the case where eviction removes the entry but a write/remove acquires the orderer and writes/removes the entry before the eviction can acquire the orderer? Or maybe it only makes a difference with passivation?

File: core/src/main/java/org/infinispan/util/concurrent/DataOperationOrderer.java [hunk 4]
```diff
@@ -35,24 +37,15 @@
...
-      while (true) {
-         current = objectStages.putIfAbsent(key, register);
-         if (current == null) {
-            break;
-         }
-         if (objectStages.replace(key, current, register)) {
-            break;
-         }
-      }
+      CompletionStage<OPERATION> current = objectStages.put(key, register);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/util/concurrent/DataOperationOrderer.java]: I forgot to nitpick about the all-caps name for `OPERATION` :)

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java [hunk 5]
```diff
@@ -540,18 +538,38 @@ public void delete(K key, InternalCacheEntry<K, V> value, RemovalCause cause) {
                   // If we are here it means this is after the eviction has completed - and thus if the
                   // entry is in the container it means the write put it there, so we skip eviction
                   if (dataContainer.containsKey(key)) {
+                     // AT this point we don't know for a fact if this WRITE has caused the entry to still be in the
+                     // DataContainer or if we are still in the eviction synchronized block just before it is removed
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: I never meant for the wording to intend it had to be in a synchronized code block, I was trying to be generic, but I can change the wording.

> I think I understood what's going on in this method, but the comments kind of focus on what's going on in this method instead of explaining why we need to do all this stuff. IIUC Caffeine evicted the entry without the orderer, so another might have acquired the orderer before the actual removal and written the entry after the removal (but PassivationCacheLoaderInterceptor.handlePassivationLoad also acquires the orderer, so reads would be blocked). So we need the 2nd containsKey to see if there was a write after the eviction/removal, in which case we don't need to passivate. We don't strictly need the first containsKey, but it allows us to avoid creating a CompletableFuture instance. Did I get it right @wburns? :)

The problem is you have two different "locks" (the actual container write lock and the orderer). Normal operations acquire the orderer then the container write lock. Eviction does the inverse order.

The main issue for this code is if you have a write to a key at the same time it is evicted. In this case the eviction has the container lock, but the write most likely has the orderer. In that case the eviction will take place removing the object and it is no where (while then writing it to the store). But as you mention reads are blocked, so you can't see that state in the cache. But then the write completes (removes nothing from the store) and the eviction finally writes to the store. Then you can end up with a value in both places.

We have two eviction implementations: Caffeine and our off heap. During eviction the former removes the entry after the listener is invoked, where as off heap removes the entry before notifying.

The problem this code fixes is that we don't for a fact know if this callback is invoked in the eviction thread or a different thread, all we know is that the write has completed. The former we hold the key lock for this container, the latter we do not (we still hold the orderer though). The delay is there to guarantee we are checking the container outside of the container lock during eviction.

Once we are outside of the container lock we can see if the entry is in the container to skip passivation, otherwise we have to passivate it, to keep it in the cache.

The extra containsKey is to avoid the lambda as you mentioned if we happen to not be in the lock any longer.

I believe Caffeine evicts in this way to prevent an eviction if the user desired, by throwing an exception.

I am updating the comments to hopefully make this more clear.

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java [hunk 6]
```diff
@@ -540,18 +538,38 @@ public void delete(K key, InternalCacheEntry<K, V> value, RemovalCause cause) {
...
+                              return skipEviction(orderer, key, future, operation);
+                           } else {
+                              return handleNotificationAndOrderer(key, entry, passivator.passivateAsync(entry), orderer, evictionManager, future);
+                           }
+                        });
+                     }
                      return skipEviction(orderer, key, future, operation);
                   }
                default:
+                  CompletionStage<Void> passivatedStage = passivator.passivateAsync(entry);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: Sure

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java [hunk 7]
```diff
@@ -572,10 +590,26 @@ public void delete(K key, InternalCacheEntry<K, V> value, RemovalCause cause) {
...
+         ensureEvictionDone.put(key, future);
+         handleEviction(value, orderer, passivator.running(), evictionManager, AbstractInternalDataContainer.this,
+               future);
+      }
+
+      // It is very important that the fact that this method is invoked AFTER the entry has been evicted outside of the
+      // lock. This way we can see if the entry has been updated concurrently with an eviction properly
+      @Override
+      public void onRemoval(K key, InternalCacheEntry<K, V> value, RemovalCause cause) {
+         if (cause == RemovalCause.SIZE) {
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: Yes, `onEntryChosenForEviction` is invoked while holding the lock for the given key. Thus we use the fact that `onRemoval` is invoked outside the lock after it has been removed.

File: core/src/test/java/org/infinispan/eviction/impl/EvictionWithConcurrentOperationsTest.java (hunk 3 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/eviction/impl/EvictionWithConcurrentOperationsTest.java]: It only matters with passivation. testWriteDuringEviction of the other class should be handling that.

**Comment by wburns** [comment]
Updated per all the comments

**Comment by danberindei** [comment]
I got a failure on my machine:
[pr_wburns_ISPN-10458_review_20190926-1956_EvictionWithPassivationAndConcurrentOperationsTest-infinispan-core.log.gz](https://github.com/infinispan/infinispan/files/3661445/pr_wburns_ISPN-10458_review_20190926-1956_EvictionWithPassivationAndConcurrentOperationsTest-infinispan-core.log.gz)

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java [hunk 8]
```diff
@@ -532,50 +532,97 @@ public void delete(K key, InternalCacheEntry<K, V> value, RemovalCause cause) {
...
+                  // 2. The thread that completed the orderer that does NOT hold the lock
+
+                  // Depending upon the underlying eviction algorithm it may be that if you are in thread 1 above
+                  // that the entry may or may not be removed while being notified. Due to this we can't tell if
+                  // the entry is actually evicted at this point or not. Thus we have an optional selfDelay that can
+                  // be passed to notify this method that the eviction is complete and can look at the container properly
+
                   if (dataContainer.containsKey(key)) {
-                     return skipEviction(orderer, key, future, operation);
+                     // If a delay is provided that means we have to wait until it completes to guarantee this code
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: I personally am not worried about that level of performance here. This should have very infrequently (requires evicting while writing the same key) and would rather not complicate it with additional conditionals.

File: core/src/main/java/org/infinispan/interceptors/distribution/ScatteredDistributionInterceptor.java [hunk 9]
```diff
@@ -141,7 +141,8 @@
    @Inject protected KeyPartitioner keyPartitioner;
    @Inject PersistenceManager persistenceManager;
    @Inject Configuration configuration;
-   @Inject NonBlockingOrderer orderer;
+   @Inject
+   DataOperationOrderer orderer;
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/distribution/ScatteredDistributionInterceptor.java]: Fixed.

**Comment by wburns** [comment]
Fixed comments, looking at failure.

**Comment by wburns** [comment]
Test failure was caused by not replacing the original orderer in the test. So both threads were stuck waiting on the checkpoint and it was non deterministic which would acquire the orderer first. Now the test guarantees the proper thread retrieves the orderer as the test was designed to do.

**Comment by wburns** [comment]
Also added a couple more trace messages just to be safe :)

**Comment by wburns** [comment]
Should be ready again @danberindei !

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: Ok. I was also looking for a way to remove the fall through to the next switch case, but that's just my OCD acting up :)

My only remaining comment is that I was trying to centralize the algorithm description in the javadoc so we could remove (most of) the inline comments in the method body.

[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: Ah, I copied over the description you had, did you want more over that?

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java [hunk 10]
```diff
@@ -498,84 +498,157 @@ public void delete(K key, InternalCacheEntry<K, V> value, RemovalCause cause) {
...
       return stage;
    }
 
-   final class DefaultEvictionListener {
+   final class DefaultEvictionListener implements RemovalListener<K, InternalCacheEntry<K, V>> {
+      Map<Object, CompletableFuture<Void>> ensureEvictionDone = new ConcurrentHashMap<>();
 
       void onEntryChosenForEviction(K key, InternalCacheEntry<K, V> value) {
-         handleEviction(key, value, orderer, passivator.running(), evictionManager, AbstractInternalDataContainer.this);
+         CompletableFuture<Void> future = new CompletableFuture<>();
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: Please add a comment here as well, e.g.

// Schedule an eviction to happen after the key lock is released

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: Yeah, it seemed to me that the inline comments overlap with each other and I thought there would be no overlap if all the comments were in one place :)

[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: So you want me to remove the comments in the method it sounds like :)

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java (hunk 10 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: Okay.

File: core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java]: Yep, unless something doesn't overlap with the javadoc

**Comment by wburns** [comment]
Updated and rebased.

**Comment by danberindei** [comment]
Integrated, thanks Will!
